File: leetcode/2.py
class Node:
    """
    A single node of a linked-list.
    """
    def __init__(self, val, next=None):
        self.val, self.next = val, next

    def append(node):
        if node is None or node is not Node:
            raise RuntimeError("node must be of type Node")

        self.next = node


def carry_add(x, y, carry=0):
    value = x + y + carry
    carry = 0

    if value >= 10:
        value -= 10
        carry = 1

    return (value, carry)


def add(l1, l2):
    """Adds two numbers represented by linked-lists
    with least significant digit at the head of the list"""
    value, carry = carry_add(l1.val, l2.val)

    result = Node(value)
    head = result # save reference for later

    while (l1.next is not None and l2.next is not None):
        l1, l2 = l1.next, l2.next
        value, carry = carry_add(l1.val, l2.val, carry)
        result.next = Node(value)
        result = result.next

    while (l1.next is None and l2.next is not None):
        l2 = l2.next
        value, carry = carry_add(l2.val, carry)
        result.next = Node(value)
        result = result.next

    while (l1.next is not None and l2.next is None):
        l1 = l1.next
        value, carry = carry_add(l1.val, carry)
        result.next = Node(value)
        result = result.next

    if carry > 0:
        result.next = Node(carry)

    return head


def print_list(l):
    while l.next is not None:
        print("{} -> ".format(l.val)),
        l = l.next

    print("{}".format(l.val))

def main():
    l1 = Node(9, Node(1, Node(6)))
    l2 = Node(0)
    result = add(l1, l2)
    print_list(result)


if __name__ == "__main__":
    main()

File: leetcode/2.c
#include<stdio.h>
#include<stdlib.h>

/*
 *
 *
 */
struct list {
    int value;
    struct list * next;
};

/*
 * Appends a node with given value to list.
 */
struct list* append(struct list* l, int value) {
    if (l == NULL) return l;
    l->next = malloc(sizeof(struct list));
    l->next->value = value;
}

/*
 * Print a list starting from the head.
 */
void print_list(struct list* l) {
    struct list* curr = l;
    while (curr->next != NULL) {
        printf("%d -> ", curr->value);
        curr = curr->next;
    }
    printf("%d\n", curr->value);
}

/*
 * Add two numbers represented as linked-lists with the
 * least-significant digit living at the head (reversed):
 *
 *   1 -> 1 -> 9 = 911
 * + 0 -> 2 -> 4 = 420
 * -------------------
 *        
*/
struct list* add(struct list* l1, struct list* l2) {
    struct list* result = malloc(sizeof(struct list));
    int value = l1->value + l2->value;
    int carry = 0;

    if (value >= 10) {
        value -= 10;
        carry = 1;
    }

    result->value = value;
    struct list* head = result; // save pointer for later

    while (l1->next != NULL && l2->next != NULL) {
        l1 = l1->next;
        l2 = l2->next;

        value = l1->value + l2->value;

        if (carry > 1) {
            value += carry;
            carry = 0;
        }

        if (value >= 10) {
            value -= 10;
            carry = 1;
        }

        result = append(result, value);
    }

    return head;
}

int main() {
    struct list * l1 = malloc(sizeof(struct list));
    struct list * l2 = malloc(sizeof(struct list));

    l1->value = 2;
    l2->value = 5;

    append(l1, 1);
    append(l1->next, 3);
    append(l2, 4);
    append(l2->next, 8);

    print_list(l1);
    print_list(l2);

    struct list* result = add(l1, l2);
    print_list(result);
}